pytest_timeout.py

```diff
@@ -507,3 +531,9 @@ def dump_stacks(terminal):
             thread_name = "<unknown>"
         terminal.sep("~", title="Stack of %s (%s)" % (thread_name, thread_ident))
         terminal.write("".join(traceback.format_stack(frame)))
+
+
+def pytest_runtest_logfinish(nodeid, location):
+    if _suite_expire_time and _suite_expire_time < time.time():
+        pytest.exit(f"suite-timeout: {_suite_timeout_minutes} minutes exceeded",
+                    returncode=0)
```

Comment:
This should not be a successful exit should it be?  Is `returncode=0` correct?

Reply:
Using this hook seems like it might ineterfere with reporting of the last test, especially since you don't use trylast or something.

Have you considered doing this in the existing `pytest_runtest_protocol` hook?  There you have access to the `session` (via `item`) and can use `session.shouldfail` which is documented (in `pytest_runtestloop` docs) to be a way to terminate the suite.

Additionally since item gives you a great deal of access, you can also get rid of the globals by storing the session timeout directly in the config.  I'm not too keen on those globals.

I'm not saying what I suggest here is right, but it needs experimentation and consideration of what the tradeoffs are to achieve this.

Reply:
I like the idea of using `pytest_runtest_protocol`, but it's difficult.  
Setting `session.shouldstop` or `session.shouldfail` makes a lot of sense. (not sure which one though).
So I want to do that.

However, I don't know how to add code "at the end" of `pytest_runtest_protocol`.
If I add the timeout using this hook, it then completes the protocol for that item. 
So, at the end of a test, if we've exceeded the timeout, we don't know it until the runtest_protocol for the next test. 
Then we find out and finish. 
So we end up running an extra test after the timeout.

The closest thing I can figure out is to use `make_report` and `shouldstop` (or `shouldfail`).
```
def pytest_runtest_makereport(item, call):
    session = item.session
    if _expire_time and _expire_time < time.time():
        session.shouldstop = f"suite-timeout: {_suite_timeout} sec exceeded"
```

I think this is probably the best of both worlds with the hooks available, and doesn't use `exit()` which is rather ugly. I actually just didn't know about `shouldfail` and `shouldstop`. Thanks for pointing those out.

Reply:
ooo. my first use of StashKey, fun.